### DISCUSSION TITLE
UpdateAsync() Causing error

### DISCUSSION THREAD

**Discussion by StretchFredrik** [pnp/pnpcore#1055] [Q&A]
Hello!

Currently im struggling with updating the title field in my solution, see code below:

![image](https://user-images.githubusercontent.com/120383942/207042779-3e80ac53-d0ad-4412-8b70-9cf6e296cecf.png)

But every time it crashes on the step "UpdateOverwriteVersionAsync" giving me the following error:

![image](https://user-images.githubusercontent.com/120383942/207042950-175dc31e-6b5c-429e-a5be-2dc95f177ce3.png)

Any ideas on whats the issue is? All help is appreciated!

**Comment by jansenbe** [comment]
@StretchFredrik : I've just tried this in my lab using a regular document library and do not see an issue. Might be that this is related to the docId feature (https://support.microsoft.com/en-us/office/enable-and-configure-unique-document-ids-ea7fee86-bd6f-4cc8-9365-8086e794c984) being active here. Can you see if you've the same issue in a regular doc lib? 

Also, did you try using the `UpdateAsync` method and see if that works?

**Reply by StretchFredrik** [reply]
Thanks for your answer! :) 

I tried using the UpdateAsync() but it gets the same error just with code 400 instead of 200.

We need the DocumentID feature active in our solution to make it work im afraid :( . Is it not possible to use a library with Document-ID service active and use the methods above?

**Reply by jansenbe** [reply]
@StretchFredrik:  I've enabled the doc id service on a site collection:
![image](https://user-images.githubusercontent.com/7451219/207052411-e14c1496-0c7d-4c42-8fd1-98a6bc4d0d22.png)

Using this test case:

![image](https://user-images.githubusercontent.com/7451219/207052380-72b17232-33da-4768-9d32-22fd3773904e.png)

still works fine. Would be interesting to see which fields you've loaded, in my case there's no dlc_DocId_Url field

![image](https://user-images.githubusercontent.com/7451219/207051802-437aaf2b-e38c-4ab1-ab9e-9fd72697f74d.png)

**Reply by StretchFredrik** [reply]
I get the following values:

![image](https://user-images.githubusercontent.com/120383942/207054301-079de313-83b0-49b5-8d28-9f242f0f4f5f.png)

**Reply by StretchFredrik** [reply]
I made a completely new site, created a library, enabled document-ID and i still get the docidUrl column. Very weird.

**Reply by jansenbe** [reply]
@StretchFredrik : interesting, if you expand "OData__dlc_DocIdUrl" what do you then see? Is the FieldUrlValue correclty set?

**Reply by jansenbe** [reply]
@StretchFredrik : also, did you try loading the file via `GetFileByServerRelativeUrlAsync(filename, , f => f.ListItemAllFields)` already?

**Reply by StretchFredrik** [reply]
This is what i get in the docidUrl column:

![image](https://user-images.githubusercontent.com/120383942/207064867-f850e5ef-da45-42b2-aca2-8b9e44d0abc0.png)

Yes my latest test looked like this:

`IFile updateFile = await webdest.GetFileByServerRelativeUrlAsync(this.siteUrl + doc.Path, f => f.ListItemAllFields);
                    await updateFile.ListItemAllFields.GetAsync();
                    updateFile.ListItemAllFields["Title"] = "Hej";
                    await updateFile.ListItemAllFields.UpdateOverwriteVersionAsync();`

Thanks for your help! :)

**Reply by jansenbe** [reply]
@StretchFredrik : can you remove `updateFile.ListItemAllFields.GetAsync();`, it's not needed and might cause the issue

**Reply by StretchFredrik** [reply]
I get the same error im afraid :( 

`ERROR: HttpResponseCode: 200
Message: Column 'OData__dlc_DocIdUrl' does not exist. It may have been deleted by another user. `

**Reply by jansenbe** [reply]
@StretchFredrik : I assume you're using version 1.8 or a 1.8 nightly? Looking at the raw API call the OData__dlc_DocIdUrl field is returned for me, but not loaded. Will investigate that first

```json
{
    "ListItemAllFields": {
        "FileSystemObjectType": 0,
        "Id": 1,
        "ServerRedirectedEmbedUri": null,
        "ServerRedirectedEmbedUrl": "",
        "ContentTypeId": "0x01010084A8024305D0504BA639615AC5743FCA",
        "ComplianceAssetId": null,
        "Title": "Test3",
        "OData__dlc_DocId": "HCDF6TP43JRV-873605472-1",
        "OData__dlc_DocIdUrl": {
            "Description": "HCDF6TP43JRV-873605472-1",
            "Url": "https://bertonline.sharepoint.com/sites/espc2022-92/_layouts/15/DocIdRedir.aspx?ID=HCDF6TP43JRV-873605472-1"
        },
        "ID": 1,
        "Created": "2022-12-12T05:00:02",
        "AuthorId": 6,
        "Modified": "2022-12-12T07:02:38",
        "EditorId": 6,
        "OData__CopySource": null,
        "CheckoutUserId": null,
        "OData__UIVersionString": "1.0",
        "GUID": "4bf26515-2d0b-4a87-89ed-9e215a4558f0"
    },
    "UniqueId": "669963c6-2d0b-4838-8c61-451f1e18de9e"
}
```

**Reply by jansenbe** [reply]
@StretchFredrik : can you try replacing `f => f.ListItemAllFields` with ` f => f.ListItemAllFields.QueryProperties(li=>li.All, 
        li => li.ParentList.QueryProperties(p => p.Title, 
            p => p.Fields.QueryProperties(p => p.InternalName, p => p.FieldTypeKind, p => p.TypeAsString, p => p.Title)))` . The issue (with the current release at least) is that when a file is loaded the code doesn't know about the parent list of the file and the parent list is needed to understand the field type (e.g. is it a URL field). By loading the parentlist as part of loading the file's listitem this is solved. A bonus is that all the needed list information (like the list fields) can be loaded in a single request.

So below code (feel free to replace `GetFileByLinkAsync` with `GetFileByServerRelativeUrlAsync` or `GetFileByIdAsync`) works for me. The OData__dlc_DocIdUrl is correctly loaded and updating works fine.

```csharp
IFile doc = await context.Web.GetFileByLinkAsync($"https://bertonline.sharepoint.com/sites/espc2022-92/Shared%20Documents/SyntexSamples.png", 
    f => f.ListItemAllFields.QueryProperties(li=>li.All, 
        li => li.ParentList.QueryProperties(p => p.Title, 
            p => p.Fields.QueryProperties(p => p.InternalName, p => p.FieldTypeKind, p => p.TypeAsString, p => p.Title))));
                
doc.ListItemAllFields["Title"] = "Test2";
await doc.ListItemAllFields.UpdateOverwriteVersionAsync();
// Using the regular list update also works fine
// await doc.ListItemAllFields.UpdateAsync()
```

Also see https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#updatingreading-file-metadata.

**Reply by StretchFredrik** [reply]
Thanks for your answer! =) 

I am using version 1.0.0, i tried updating to 1.8 but got a bunch of errors in the following code: 

` d => d.Files,
            d => d.Folders,
            d => d.Folders.LoadProperties(
                d => d.Name,
                d => d.UniqueId),
            d => d.Files.LoadProperties(
                d => d.ServerRelativeUrl,
                d => d.MajorVersion,
                d => d.ListItemAllFields,
                d => d.ListItemAllFields.LoadProperties(
                    d => d.Id,
                    d => d.Values,
                    d => d.Values)) };`

Where can i find patchnotes to see what the LoadProperties should be replaced with in my case after update?

Maybe this is just a issue in the older version im using? I cannot try your code since QueryProperties isnt accepted as a method in my version.

**Reply by jansenbe** [reply]
@StretchFredrik : please ensure you've the PnP.Core.Model namespace added and use `QueryProperties`. Also 1.0 is very old and we don't do fixes in older releases, so strongly advice to use the most recent version.

**Reply by StretchFredrik** [reply]
My old "LoadProperties" is that replaced by QueryProperties?

Where can i find the information about the different methods? Currently the solution is complaining that GetAsync doesn't exist anymore after the update =)

**Reply by jansenbe** [reply]
@StretchFredrik : seems you've been using an even older version then 1.0.0?? See https://pnp.github.io/pnpcore/using-the-sdk/upgrade-to-beta3.html for the key changes we did going from beta2 to later releases.

**Reply by StretchFredrik** [reply]
Thank you @jansenbe  i now got it working with the latest version, but im not sure how the CopyToAsync works. Every time i use it it deletes the current file if there is one, and creates it as a new one instead of replacing the current one. This makes the document get a new ID and also a new link, which is not optimal.  Any ideas why?

**Reply by jansenbe** [reply]
@StretchFredrik : what options are you providing for that method? Also are you copying within the same web or to another site collection or another web? Also, see https://pnp.github.io/pnpcore/using-the-sdk/files-intro.html#copying-and-moving-files for relevant docs

**Reply by StretchFredrik** [reply]
Using the following settings:

`await file.CopyToAsync(this.siteFullUrl + doc.Path, true , new PnP.Core.Model.SharePoint.MoveCopyOptions()
                        {
                            ShouldBypassSharedLocks = true,
                            ResetAuthorAndCreatedOnCopy = false,
                            KeepBoth = false
                        });`

I am copying to a different site. So copy from Site A to Site B
